src/Datagrid/Datagrid.php

```diff
@@ -182,11 +235,61 @@ public function bind(AdminInterface $admin, Request $request)
         $this->configureOptions($resolver);
         $this->options = $resolver->resolve($admin->getDatagridOptions());
 
-        // attach our fields, filters and actions
-        $admin->buildDatagrid($this);
-
         $this->admin = $admin;
         $this->request = $request;
+
+        // attach our fields, filters and actions
+        $this->admin->buildDatagrid($this);
+
+        $this->applyFilters();
+    }
+
+    /**
+     * Prepare filters form, bind filter values to filters
+     */
+    private function applyFilters()
+    {
+        // Manage filters
+        // Filter values come either from the form (GET)
+        // Or from session
+        $this->buildFiltersForm();
+
+        if (!$this->filtersForm) {
+            // No filters are configured for this admin
+            return;
+        }
+
+        $this->retrieveFilterValues();
+        // Init form with current data if any
+        $this->filtersForm->setData($this->filterValues);
+
+        $this->filtersForm->handleRequest($this->request);
+
+        if ($this->filtersForm->isSubmitted()) {
+
+            if ($this->filtersForm->get('reset')->isClicked()) {
+                // Submitted for reset
+                // Reset stored content
+                $this->clearFilterValues();
+                // Reset displayed values by rebuilding the form
+                // yes, it's a pity we cannot simply change content of it
+                $this->buildFiltersForm();
+            }
+            else {
+                // Submitted for filtering
+                // Overwrite any stored values with the submitted ones
+                $this->filterValues = $this->filtersForm->getData();
+                // Update storage for subsequent requests
+                $this->storeFilterValues();
```

Comment:
should probably be this->storeFilterValues($data) instead of setting the values implicitly the line above.